Fix setting and updating size of transient's frame

tarsius opened the pull request:
This fixes the regression described in #5.  Additionally fixes a second issue, which (I assume) has always existed.  Please see 4db747a8d1a332c3b634cf4c302c14c9c96bf01f for details.

#6 also tries to fix the regression, but it just hacks around it without understanding the cause.

I've also included some cleanup and minor improvements.  Feel free to drop the parts you don't like.

aaronjensen commented:
This is much improved as it actually sizes properly. One thing I noticed is that the bottom has an extra gray border that's visible. I'm not sure if it's intentional or not or maybe something in my config. It looks like the color of my mode line, so maybe it's that?

![CleanShot 2024-11-10 at 18 06 30@2x](https://github.com/user-attachments/assets/f96e7bb6-6401-41b4-83d9-9f66cb97302a)

tarsius commented:
That line (or rather its color) serves a purpose, but it's not essential (most users don't know about it and honestly I don't pay attention to it either).

Customize `transient-mode-line-format` to hide it, change its thickness, or display a mode-line instead.

aaronjensen commented:
Thank you.

seagle0128 commented:
Kindly ping. Can it be merged?

seagle0128 commented:
> This is much improved as it actually sizes properly. One thing I noticed is that the bottom has an extra gray border that's visible. I'm not sure if it's intentional or not or maybe something in my config. It looks like the color of my mode line, so maybe it's that?

There is a trick to remove the gray line: adding this to the parameter of `posframe-show`.

```elisp
   :height (with-current-buffer buffer (1- (count-screen-lines (point-min) (point-max))))
```

seagle0128 commented:
@tarsius I checked the source code, and there is no `transient--fit-window-to-buffer` in the original file. Can you please double check?

aaronjensen commented:
@seagle0128 
> There is a trick to remove the gray line...

As dgutov mentioned, the most direct way to remove it is `(setq transient-mode-line-format nil)`.

Regarding `transient--fit-window-to-buffer`, please make sure you are using the version from GitHub, rather than the one bundled with Emacs.

seagle0128 commented:
@aaronjensen Thank you for letting me know.

tarsius commented:
I've made some changes in Transient to better support displaying the transient menu in a dedicated frame. These will be releases in v0.8.2, very early next year.

I've also updated this pr to take advantage of these improvements. (And sneaked in a some unrelated fixes/changes.)

This package now boils down to:

1. `transient-posframe-mode` now does nothing but switch between two values for `transient-display-action`, the posframe value and the other value. It no longer has to [de-]activate any advice, since no advice is needed anymore.
2. `transient-posframe--show-buffer`, the display action function, basically just calls `posframe-show` with the appropriate arguments.
3. A bunch of options to allow tuning the arguments to `posframe-show`.

The mode (1) isn't really necessary if you want to always use posframe. I also feel that by using options (3) we just add another layer, which reduces flexibility and adds complexity. The display action function (2) is really all we need, and users can copy-paste and then edit it directly to get back full control.

I.e., I recommend you use this snippet, instead of this package (at least until the maintainer merges this and makes further improvements):

```elisp
(setq transient-mode-line-format nil)
(setq transient-display-buffer-action
      (list
       (lambda (buffer _)
         (posframe-show
          buffer
          :poshandler #'posframe-poshandler-frame-center
          :min-width transient-minimal-frame-width
          :lines-truncate t
          :internal-border-color (transient--prefix-color)
          :internal-border-width 1)
         (get-buffer-window transient--buffer t))))
```

Or with comments:

```elisp
;; By default a thin line whose color indicates the transient-ness of the
;; menu is used.  Without an echo area that would look odd and below we
;; color the border instead.
(setq transient-mode-line-format nil)

(setq transient-display-buffer-action
      (list
       (lambda (buffer _)
         (posframe-show
          buffer
          :poshandler #'posframe-poshandler-frame-center
          ;; To reduce the likelyhood of horizontal resizing, use the
          ;; same minimal width as transient uses by default.  It matches
          ;; the width needed to display the commands common to all menus.
          :min-width transient-minimal-frame-width
          ;; If the parent frame is small, there might not be enough room.
          ;; By default posframe wraps lines, but we truncate instead.
          :lines-truncate t
          ;; Enable the fringe, so that we can see when truncation has
          ;; occured.  Hm, actually that's not good enough, so let's not.
          ;; :right-fringe 8
          ;;
          ;; Indicate transient-ness of the menu.  You could also use a
          ;; constant color, if you don't care about this.
          :internal-border-color (transient--prefix-color)
          :internal-border-width 1)
         ;; `posframe-show' it not suitable for use as a display action
         ;; and it appears posframe does not provide some other function
         ;; that is.  We can make this more complient by at least
         ;; returning the used window.
         (get-buffer-window transient--buffer t))))
```

This comes with some improvements that you currently cannot achieve by tweaking `transient-posframe`'s options.

By the way, this is how you could use a non-child frame:

```elisp
(setq transient-display-buffer-action
      '(display-buffer-pop-up-frame
        ;; For some window managers this might be enough.  For i3 it is
        ;; not, unless we add the following, which would have unintended
        ;; consequences.
        ;;   focus_follows_mouse no
        ;;   focus_on_window_activation none
        (inhibit-switch-frame . t)
        ;; This however gets the job done, at least for i3.  It causes
        ;; the previously selected xwindow to be re-selected.  Note that
        ;; setting this to nil, is not the same as leaving this undefined.
        (post-command-select-window . nil)
        ;; The presense of this causes `transient--display-buffer' to
        ;; pre-calculate the approximate size of the frame, which reduces
        ;; flickering and improves placement.  The exact size is also
        ;; automatically set (after creating the frame), so it would be
        ;; reduncant/uneffective to specify the sizes here (by setting
        ;; `height' and `width').
        (pop-up-frame-parameters
         ;; Of course i3 has be told to actually float our xwindow.
         ;;   for_window [title="transient--buffer"] floating enable, border pixel 1
         ;;   no_focus [title="transient--buffer"]
         (name . "transient--buffer")
         ;; There's not really a need for fringes here.
         (left-fringe . 0)
         (right-fringe . 0))
        ;; This resizing would still happen after the frame is created.
        ;; This is effectively the same as doing it in transient--show,
        ;; which we already do.
        ;; (window-size . transient--fit-window-to-buffer)
        ))
```

I have not yet looked into using a child-frame, but *without* using posframe.

tarsius commented:
I've implemented similar fixes for `which-key-posframe`. I did use that myself at the time, but eventually I started using `posframe-show` directly, as I suggest you do for this package (see above).

I just noticed that eventually I forked `which-key-posframe` to the Emacsorphanage and changed Melpa to get it from there https://github.com/yanghaoxie/which-key-posframe/issues/19.

I do care about `transient-posframe` not being broken, because as the maintainer of `transient` I am affected when users run into issues and possibly end up concluding that the issue is with `transient` itself. So I have decided to move my fork to the Emacsorphanage and to change Melpa to get it from there.

https://github.com/emacsorphanage/transient-posframe

I will gladly revert this on Melpa and Emacsmirror once the package is being maintained here again.

seagle0128 commented:
> https://github.com/emacsorphanage/transient-posframe
> 

I can confirm it's working well. Thanks, @tarsius

tarsius commented:
I am tidying up my list off pending pull-requests, but this still holds true:

> I will gladly revert this on Melpa and Emacsmirror once the package is being maintained here again.
